python/ql/src/experimental/semmle/python/frameworks/Fabric.qll

```diff
@@ -0,0 +1,578 @@
+      }
+
+      /**
+       * Gets a reference to the attribute `attr_name` of the `fabric.group` module.
+       * WARNING: Only holds for a few predefined attributes.
+       */
+      private DataFlow::Node group_attr(string attr_name) {
+        result = group_attr(DataFlow::TypeTracker::end(), attr_name)
+      }
+
+      /**
+       * Provides models for the `fabric.group.Group` class and its subclasses.
+       *
+       * `fabric.group.Group` is an abstract class, that has concrete implementations
+       * `SerialGroup` and `ThreadingGroup`.
+       *
+       * See
+       * - https://docs.fabfile.org/en/2.5/api/group.html#fabric.group.Group
+       * - https://docs.fabfile.org/en/2.5/api/group.html#fabric.group.SerialGroup
+       * - https://docs.fabfile.org/en/2.5/api/group.html#fabric.group.ThreadingGroup
+       */
+      module Group {
+        /**
+         * A source of an instance of a subclass of `fabric.group.Group`
+         *
+         * This can include instantiation of a class, return value from function
+         * calls, or a special parameter that will be set when functions are call by external
+         * library.
+         *
+         * Use `Group::subclassInstance()` predicate to get references to an instance of a subclass of `fabric.group.Group`.
+         */
+        abstract class SubclassInstanceSource extends DataFlow::Node { }
+
+        /** Gets a reference to an instance of a subclass of `fabric.group.Group`. */
+        private DataFlow::Node subclassInstance(DataFlow::TypeTracker t) {
+          t.start() and
+          result instanceof SubclassInstanceSource
+          or
+          exists(DataFlow::TypeTracker t2 | result = subclassInstance(t2).track(t2, t))
+        }
+
+        /** Gets a reference to an instance of a subclass of `fabric.group.Group`. */
+        DataFlow::Node subclassInstance() {
+          result = subclassInstance(DataFlow::TypeTracker::end())
+        }
+
+        /**
+         * Gets a reference to the `run` method on an instance of a subclass of `fabric.group.Group`.
+         *
+         * See https://docs.fabfile.org/en/2.5/api/group.html#fabric.group.Group.run
+         */
+        private DataFlow::Node subclassInstanceRunMethod(DataFlow::TypeTracker t) {
+          t.startInAttr("run") and
+          result = subclassInstance()
+          or
+          exists(DataFlow::TypeTracker t2 | result = subclassInstanceRunMethod(t2).track(t2, t))
+        }
+
+        /**
+         * Gets a reference to the `run` method on an instance of a subclass of `fabric.group.Group`.
+         *
+         * See https://docs.fabfile.org/en/2.5/api/group.html#fabric.group.Group.run
+         */
+        DataFlow::Node subclassInstanceRunMethod() {
+          result = subclassInstanceRunMethod(DataFlow::TypeTracker::end())
+        }
+      }
+
+      /**
+       * A call to `run`, `sudo` on an instance of a subclass of `fabric.group.Group`.
```

Comment:
How is `sudo` tracked here?

Reply:
too much copy paste :neutral_face: there is no `sudo` method on a Group, so I'll just remove that bit :+1: